scripts/base/frameworks/storage/main.zeek

```diff
@@ -0,0 +1,170 @@
+
+		# The script-level type of keys stored in the backend. Used for
+		# validation of keys passed to other framework methods.
+		key: any;
+
+		# The script-level type of keys stored in the backend. Used for
+		# validation of values passed to ``put`` as well for type
+		# conversions for return values from ``get``.
+		value: any;
+
+		# Indicates whether this value should overwrite an existing entry
+		# for the key.
+		overwrite: bool &default=F;
+
+		# An interval of time until the entry is automatically removed from the
+		# backend.
+		expire_time: interval &default=0sec;
+
+		# Indicates whether this operation should happen
+		# asynchronously. If this is true, the call to put must happen
+		# as part of a when statement. This flag is overridden when
+		# reading pcaps, since time won't move forward the same as when
+		# caputring live traffic.
+		async_mode: bool &default=T;
+	};
+
+	## Options record for the built-in SQLite backend.
+	type SqliteOptions: record {
+		## Path to the database file on disk
+		database_path: string;
+
+		## Name of the table used for storing data
+		table_name: string;
+
+		## Key/value table for passing tuning parameters when opening
+		## the database.  These must be pairs that can be passed to the
+		## ``pragma`` command in sqlite.
+		tuning_params: table[string] of string &default=table(
+			["journal_mode"] = "WAL",
+			["synchronous"] = "normal",
+			["temp_store"] = "memory"
+		);
+	};
+
+	## Options record for the built-in Redis backend.
+	type RedisOptions: record {
+		# Address to the server
+		server_addr: string;
+
+		# Port for the server
+		server_port: port;
+
+		# Server unix socket file. This can be used instead of the
+		# address and port above to connect to a local server.
+		server_unix_socket: string;
+
+		# Prefix used in key values stored to differentiate varying
+		# types of data on the same server.
+		key_prefix: string;
+
+		# Redis only supports sync and async separately. You cannot do
+		# both with the same connection. If this flag is true, the
+		# connection will be async and will only allow commands via
+		# ``when`` commands. You will still need to set the
+		# ``async_mode`` flags of the put, get, and erase methods to
+		# match this flag. This flag is overridden when reading pcaps
+		# and the backend will be forced into synchronous mode, since
+		# time won't move forward the same as when caputring live
+		# traffic.
+		async_mode: bool &default=T;
+	};
+
+	## Opens a new backend connection based on a configuration object.
+	##
+	## btype: A tag indicating what type of backend should be opened.
+	##
+	## config: A record containing the configuration for the connection.
+	##
+	## key_type: The Val type of the key being stored.
+	##
+	## val_type: The Val type of the key being stored.
+	##
+	## Returns: A handle to the new backend connection, or null if the connection failed.
+	global open_backend: function(btype: Storage::Backend, config: any, key_type: any,
```

Comment:
Since both `open_backend` and `close_backend` block they should probably be only called during initialization. Do you think we might have a need for async versions for some (future) backends? In that case it might make migrate from global open/close methods to backend-specific versions.

Reply:
I'm trying to think of a case where you would ever open a backend outside of initialization. If `open_backend` was allowed to be async, the backend handle could only live during the context of a `when` statement since the handle would be invalid otherwise until the trigger happened.

Reply:
@bbannier and I talked through this a bit outside of the PR. Building in async for `open_backend` but leaving it sync-by-default doesn't lose anything, but leaves the option open.

Reply:
Added support for this in https://github.com/zeek/zeek/pull/4163/commits/8ecd559b72ea856b5c8caa63c8d6976a7d8b9177 and https://github.com/zeek/zeek/pull/4163/commits/3b8312d381c32118a981c71484614ba0e9b726e3